### ISSUE TITLE
list of subclasses cannot be assigned to list of superclass

### ISSUE THREAD

**Issue by sarisia** [microsoft/pyright#130]
**Describe the bug**
I have a `Base` class and a `Child` class derived from `Base` class.
When I try to assign a list of `Child` instances to `List[Base]`,
the list of `Child` is recognized as `List[Child]` and get warning.

**To Reproduce**
Reproducible code:
```python
from typing import List

class Base():
    pass

class Child(Base):
    pass

class AnotherChild(Base):
    pass


success: List[Base] = [Child(), AnotherChild()]

fail: List[Base] = [Child(), Child()]
fail2: List[Base] = [AnotherChild(), AnotherChild()]
```

In the code above, I got
```
Expression of type 'list[Child]' cannot be assigned to declared type 'List[Base]'pyright
```
for `fail` and
```
Expression of type 'list[AnotherChild]' cannot be assigned to declared type 'List[Base]'pyright
```
for `fail2`.

**Expected behavior**
Both `Child` and `AnotherChild` is inherited from `Base` so
`List[Child]` and `List[AnotherChild]` should be able to be assigned to `List[Base]`
without warnings.

**Screenshots or Code**
![pyrightissue](https://user-images.githubusercontent.com/33576079/58050603-c6a64700-7b8a-11e9-893a-b3d1b70be3a9.jpg)

**VS Code extension or command-line**
pyright as a VS Code extension
Pyright 1.0.29

**Additional context**

**Comment by erictraut** [comment]
As non-obvious as this might sound, this is actually the correct behavior. That's because "list" and other mutable container types are considered "invariant", so the contained type needs to match exactly. If you use an immutable type like Sequence, your code will work as expected. Immutable container types are considered "covariant", so they accept subclasses.

```
success1: Sequence[Base] = [Child(), AnotherChild()]
success2: Sequence[Base] = [Child(), Child()]
success3: Sequence[Base] = [AnotherChild(), AnotherChild()]
```

If you want to understand the theory behind this, check out [PEP 483](https://www.python.org/dev/peps/pep-0483/#covariance-and-contravariance).

You might wonder why the first assignment in your example (the one that assigns the variable called `success`) works without generating an error. That's because pyright, by default, infers that list expressions that contain mixed types should be treated as List[Any]. If you enable the configuration switch `strictListInference` in your pyrightconfig.json file -- or optionally add a `# pyright: strict` comment in your source file, you'll see that an error is generated on this line as well. In strict mode, pyright infers a union type (in the case of your example, `List[Union[Child, AnotherChild]]`).

**Comment by sarisia** [comment]
Got it,
sorry for my lack of understanding for python and
thank you for your intelligible explanation
